Added CreateThread endpoint

ziad-ashraf7 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full thread system: create, view (by id/user), answer, delete, paginated listing, personalized feed.
  * Save/unsave threads and view saved items.
  * Like/unlike threads and view likes.
  * Commenting on threads with paginated comments.
  * Anonymous questions and new thread statuses (Answered, Pending).

* **Bug Fixes**
  * Fixed re-like behavior to reliably update timestamps and like counts.

* **Chores**
  * Adjusted AutoMapper and Swagger UI package versions; added paging support.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ziad-ashraf7 committed:
Added CreateThread endpoint

ziad-ashraf7 commented:
@coderabbit review

ziad-ashraf7 committed:
Feat: Added the answer thread endpoint, created the ThreadLikeController

ziad-ashraf7 commented:
@coderabbit review

ziad-ashraf7 committed:
Finished the ThreadLikeService and endpoints

ziad-ashraf7 commented:
@coderabbit review

ziad-ashraf7 commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,188 @@
+            await _unitOfWork.Threads.UpdateAsync(thread);
+            await _unitOfWork.SaveAsync();
+
+            ThreadResponseDto threadRes = new ThreadResponseDto()
+            {
+                Id = thread.Id,
+                QuestionContent = thread.QuestionContent,
+                IsAnonymous = thread.isAnonymous,
+                CreatedAt = thread.CreatedAt,
+                AskedId = thread.AskedId,
+                AskedName = thread?.Asked.Name ?? "Unknown",
+                AskerId = thread.AskerId.Value,
+                AskerName = thread.isAnonymous? "Unknown" : thread?.Asker.Name,
+                answer = thread.AnswerContent,
+                Status = thread.Status,
```
is there any better way to write the ThreadResponseDto ??
what do you think about the Anonymous checking ?

ziad-ashraf7 committed:
Refactor: resolved most of the notes

ziad-ashraf7 committed:
Finished the Threads , implemented the Saved endpoints and modified the rest

ziad-ashraf7 commented:
@coderabbit review

ziad-ashraf7 committed:
Finish the Comment Controller and service section

ziad-ashraf7 commented:
@coderabbit review

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            // check if the Asked user exite or not
+            var askedId = createThreadDto.AskedId;
+
+            var askedUser = await _unitOfWork.Users.GetByIdAsync(askedId);
```
you have to check here if the askedUser is not deleted or not and the same with askerUser
```csharp
var askedUser = await _unitOfWork.Users.FindAsync(
            u => u.Id == createThreadDto.AskedId && !u.IsDeleted);
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                isAnonymous = createThreadDto.isAnonymous,
+                CreatedAt = DateTime.UtcNow,
+            };
+
```
Status should always be Pending: The DTO allows setting any status, but a new question should always start as Pending

```csharp
// Create thread - FORCE status to Pending  
		Thread thread = new Thread  
		{  
		    AskedId = createThreadDto.AskedId,  
		    AskerId = askerId,  
		    QuestionContent = createThreadDto.QuestionContent.Trim(),  
		    AnswerContent = string.Empty,  
		    Status = ThreadStatus.Pending, // Always pending for new threads  
		    isAnonymous = createThreadDto.isAnonymous,  
		    CreatedAt = DateTime.UtcNow,  
		    UpdatedAt = DateTime.UtcNow,  
		    IsDeleted = false  
		};
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            await _unitOfWork.Users.UpdateAsync(askedUser);
+            await _unitOfWork.Users.UpdateAsync(askerUser);
+            await _unitOfWork.SaveAsync();
+
```
why you use `UpdateAsync` twice ?? i see that you should remove both of them

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                IsAnonymous = thread.isAnonymous,
+                CreatedAt = thread.CreatedAt,
+                AskerId = thread.AskerId.Value,
+                AskerName = thread.Asker?.Name ?? "Unknown",
```
Use already-loaded entity to load Asker and Asked name
```csharp
var responseDto = new ThreadResponseDto
            {
                Id = thread.Id,
                QuestionContent = thread.QuestionContent,
                Status = thread.Status,
                IsAnonymous = thread.isAnonymous,
                CreatedAt = thread.CreatedAt,
                AskerId = thread.AskerId.Value,
                AskerName = thread.isAnonymous ? "Anonymous" : askerUser.Name, // Use already-loaded entity
                AskedId = thread.AskedId,
                AskedName = askedUser.Name // Use already-loaded entity
            };
```

0xmahmoudd commented on AskFm/AskFm.BLL/DTO/CreateThreadDto.cs:
```diff
@@ -0,0 +1,16 @@
+    
+    public bool isAnonymous { get; set; }
+    
+}
```
- you should remove Status, the user should not set the status, the service will do this.
- if he creates new thread then the status will be Pending, if he answers threads then the status will be Answered
- maybe you can add some validations for improvements

```csharp
public class CreateThreadDto
{
    [Required(ErrorMessage = "Asked user ID is required")]
    public int AskedId { get; set; }

    [Required(ErrorMessage = "Question content is required")]
    [StringLength(1000, MinimumLength = 2, ErrorMessage = "Question must be between 2 and 1000 characters")]
    public string QuestionContent { get; set; }

    public bool IsAnonymous { get; set; } // Fixed naming
}
```

0xmahmoudd commented on AskFm/AskFm.BLL/DTO/AnswerThreadDto.cs:
```diff
@@ -0,0 +1,6 @@
+public class AnswerThreadDto
+{
+    public string AnswerContent { get; set; }
+}
```
you may add here some validations:
```csharp
public class AnswerThreadDto
{
    [Required(ErrorMessage = "Answer content is required")]
    [StringLength(5000, MinimumLength = 1, ErrorMessage = "Answer must be between 1 and 5000 characters")]
    public string AnswerContent { get; set; }
}
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+        var transaction = await _unitOfWork.BeginTransactionAsync();
+
+        try
+        {
```
you should first validate answer content like if empty space or more than 5000 chars
```csharp
// --- Validate answer content
            if (string.IsNullOrWhiteSpace(answerDto.AnswerContent))
            {
                await transaction.RollbackAsync();
                return await ServiceResult<ThreadResponseDto>.Failure(
                    new List<string> { "Answer cannot be empty or contain only whitespace" });
            }

            var trimmedAnswer = answerDto.AnswerContent.Trim();
            if (trimmedAnswer.Length > 5000)
            {
                await transaction.RollbackAsync();
                return await ServiceResult<ThreadResponseDto>.Failure(
                    new List<string> { "Answer cannot exceed 5000 characters" });
            }
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            var thread = await _unitOfWork.Threads.FindAsync(
+                predicate: t => t.Id == threadId,
+                includes: new[] { "Asker", "Asked" }
+            );
```
you have to check soft deleted and the status if it's pending or not 
and Update includes to add Counts for likes and comments
```csharp
// ---- Fetch thread with soft-delete and status checks
            var thread = await _unitOfWork.Threads.FindAsync(
    predicate: t => t.Id == threadId && !t.IsDeleted && t.Status == ThreadStatus.Pending,
    includes: new[] { "Asker", "Asked", "ThreadLikes", "Comments" } // Add counts
);
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            thread.AnswerContent = answerDto.AnswerContent;
+            thread.Status = ThreadStatus.Answered;
+
+            await _unitOfWork.Threads.UpdateAsync(thread);
```
you have to update DateTime also 
```csharp
// ---- update answer
            thread.AnswerContent = trimmedAnswer;
            thread.Status = ThreadStatus.Answered;
            thread.UpdatedAt = DateTime.UtcNow;

            await _unitOfWork.Threads.UpdateAsync(thread);

            // ---- here we have to add notifications before SaveAsync but this is another task
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                AskerName = thread.isAnonymous ? "Anonymous" : thread.Asker?.Name,
+                AskedId = thread.AskedId,
+                AskedName = thread.Asked?.Name
+            };
```
you have to Add `UpdateAt` , `LikesCount` and `CommentsCount`  to `threadDto`
```csharp
var threadDto = new ThreadResponseDto
{
    Id = thread.Id,
    QuestionContent = thread.QuestionContent,
    AnswerContent = thread.AnswerContent,
    Status = thread.Status,
    IsAnonymous = thread.isAnonymous,
    CreatedAt = thread.CreatedAt,
    UpdatedAt = thread.UpdatedAt, // Add this
    AskerId = thread.AskerId ?? 0,
    AskerName = thread.isAnonymous ? "Anonymous" : thread.Asker?.Name,
    AskedId = thread.AskedId,
    AskedName = thread.Asked?.Name,
    LikesCount = thread.ThreadLikes?.Count ?? 0, // Add this
    CommentsCount = thread.Comments?.Count ?? 0 // Add this
};
```

0xmahmoudd commented on AskFm/AskFm.BLL/DTO/ThreadResponseDto.cs:
```diff
@@ -0,0 +1,20 @@
+    public string? AnswerContent { get; set; }
+    public ThreadStatus Status { get; set; }
+    public bool IsAnonymous { get; set; }
+    public DateTime CreatedAt { get; set; }
```
add `UpdatedAt` attribute
`public DateTime UpdatedAt { get; set; }`

0xmahmoudd commented on AskFm/AskFm.API/Controllers/ThreadController.cs:
```diff
@@ -0,0 +1,213 @@
+    [HttpPut]
+    [Route("threads/{id}/answer")]
+    public async Task<IActionResult> AnswerQuestion([FromRoute] int id, [FromBody] AnswerThreadDto answerDto)
+    {
```
You may add here ModelState Validation
```csharp
if (!ModelState.IsValid) return BadRequest(ModelState);
```

0xmahmoudd commented on AskFm/AskFm.API/Controllers/ThreadController.cs:
```diff
@@ -0,0 +1,213 @@
+        var user = await _userService.GetCurrentUserAsync();
+        if (!user.success)
+        {
+            return BadRequest(user.Errors);
```
i think here it's better to make it `Unauthorized`
```csharp
if (!user.success)
        {
            return Unauthorized(user.Errors);
        }
```

ziad-ashraf7 replied:
but what if the problem wasn't on the authority of the user it self , what if it was an error on the `GetCurrentUserAsync()` function and the user is already authorized successfully and can do some other operations and use other endpoints that doesn't need the currentUser

Do we here still need the  `Unauthorized` return ??

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+    {
+        try
+        {
+            // Get all threads for user
```
Validate user exists and add soft-delete filtering 
```csharp
// Validate user exists
            var userExists = await _unitOfWork.Users.FindAsync(
                u => u.Id == askedId && !u.IsDeleted);
            
            if (userExists == null)
                return await ServiceResult<List<ThreadResponseDto>>.Failure(
                    new List<string> { "User not found" });

            // Add soft-delete filtering
            var threads = await _unitOfWork.Threads.FindAllAsync(
                predicate: t => t.AskedId == askedId && !t.IsDeleted,
                includes: new[] { "Asker", "Asked", "Comments", "ThreadLikes" }
            );
```

hadeer-r replied:
We added a global query filter to filter all deleted rows, so there is no need to check it again
https://github.com/Askfm-clone/Backend/blob/28cc5560d10a9c89c7a6bf76163bb06e15d85938/AskFm/AskFm.DAL/AppDbContext.cs#L53

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            if (threads == null || !threads.Any())
+            {
+                return await ServiceResult<List<ThreadResponseDto>>.Success(new List<ThreadResponseDto>());
+            }
```
You may add ordering by Descending
```csharp
threads = threads.OrderByDescending(t => t.CreatedAt).ToList();
```

0xmahmoudd commented on AskFm/AskFm.API/Controllers/ThreadController.cs:
```diff
@@ -0,0 +1,213 @@
+    [HttpGet]
+    [Route("threads")]
+    public async Task<IActionResult> GetThreads([FromQuery] int page = 1, [FromQuery] int pageSize = 10)
+    {
```
add validations for pagination 
```csharp
if (page < 1) return BadRequest(new { message = "Page must be >= 1" });

if (pageSize < 1 || pageSize > 100) return BadRequest(new { message = "PageSize must be between 1 and 100" });
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+        {
+            int skipCount = (page - 1) * pageSize;
+
+            var totalCount = await _unitOfWork.Threads.CountAsync();
```
here you have to validate that the thread is not deleted
```csharp
// Count only answered, non-deleted threads
var totalCount = await _unitOfWork.Threads.CountAsync(t => !t.IsDeleted && t.Status == ThreadStatus.Answered);
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                pageSize,
+                t => t.CreatedAt,
+                false,
+                t => true,
```
fix this
```csharp
t => !t.IsDeleted && t.Status == ThreadStatus.Answered
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                Status = thread.Status,
+                IsAnonymous = thread.isAnonymous,
+                CreatedAt = thread.CreatedAt,
+                AskerId = thread.AskerId ?? 0,
```
hide if Anonymous
`AskerId = thread.isAnonymous ? null : thread.AskerId`

0xmahmoudd commented on AskFm/AskFm.API/Controllers/ThreadController.cs:
```diff
@@ -0,0 +1,213 @@
+    [HttpGet]
+    [Route("threads/{id}")]
+    public async Task<IActionResult> GetThreadWithId([FromRoute] int id)
+    {
```
add validation for id 
```csharp
if (id <= 0) return BadRequest(new { message = "Invalid thread ID" });
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+        try
+        {
+            var thread = await _unitOfWork.Threads.FindAsync(
+                predicate: t => t.Id == id,
```
add soft-deleted check
```csharp
predicate: t => t.Id == id && !t.IsDeleted,
```

0xmahmoudd commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+                AnswerContent = thread.AnswerContent,
+                Status = thread.Status,
+                IsAnonymous = thread.isAnonymous,
+                CreatedAt = thread.CreatedAt,
```
add UodatedAt
```csharp
UpdatedAt = thread.UpdatedAt,
```

Mahmoud-Ayman-Saleh reviewed (changes requested):
**i've reviewed the next endpoints:**
- AddThread
- GetThreadById
- GetAllThreads
- AnswerThread
- GetThreads
@hadeer-r if you have a time please review the remaining endpoints, most of them needs just a validation for soft-delete and other minor validations or modifications

hadeer-r commented:
@Mahmoud-Ayman-Saleh Thanks, I will do it

hadeer-r commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+            }
+
+            // Check if user is authorized to delete this thread (either asker or asked)
+            if (thread.AskerId != userId && thread.AskedId != userId)
```
it should be `||` not `&`

ziad-ashraf7 replied:
if the `userId` is not equal to `Asker.Id` AND the `userId` is not equal to the `AskedId` , so this `userId` is not allowed to delete this thread, so I need to rollBack and return a `Failure`.

but what if the `userId` is equal to any one of them?
so the if statement won't be equal to true , and I'll continue to my logic.

but when using `||` and if the `userId != AskerId` the if statement will equal to true and it will return and `Unauthorized Failure` EVEN if the `userId was equal to AskedId`

so I think `&&` is correct.


got it ?

hadeer-r commented on AskFm/AskFm.API/Controllers/ThreadController.cs:
```diff
@@ -0,0 +1,213 @@
+        var user = await _userService.GetCurrentUserAsync();
+        if (!user.success)
+        {
+            return BadRequest(user.Errors);
```
I think `Unauthorized` is better here, as it means there is no logged in user

hadeer-r commented on AskFm/AskFm.BLL/Services/ThreadService.cs:
```diff
@@ -0,0 +1,531 @@
+
+            followedUserIds.Add(userId);
+
+            var totalCount = await _unitOfWork.Threads.CountAsync(t =>
```
There is no need for the totalCount of Threads in Feed. Also, it's an overhead on the database .
The user in the feed page doesn't care about how many answered questions there, just like on Facebook, you don't care how many of the posts you have in the feed page

ziad-ashraf7 replied:
The reason of the `totalCount` is for the pagination.
Facebook doesn't show you all the recommended posts at once, instead: it shows you some number of recommended post , and then when you scroll down , you see it loading other number of posts.

hadeer-r reviewed (changes requested):
Great work! I've reviewed the SaveThread endpoints and the ThreadLike Service and Controller.

I suggest validation attributes directly to the DTOs (for required fields and other validation logic). This will help us avoid redundant validation code in the service layer and keep things DRY.

ziad-ashraf7 committed:
refactor(totalCount & totalPages):
    - removed the totalCount and totalPages from pagedResponseDto as
      thier uses is so overhead on the database.
    - instead, replaced them with HasMore, meaning are there any
      remaining pages left here to make the client or the frontend use
      as a flage.

hadeer-r reviewed (approved):
Great work, thanks.
